src/main/java/org/apache/commons/dbcp2/managed/ManagedConnection.java

```diff
@@ -232,36 +253,56 @@ protected void transactionComplete() {
 
     @Override
     public void setAutoCommit(final boolean autoCommit) throws SQLException {
-        if (transactionContext != null) {
-            throw new SQLException("Auto-commit can not be set while enrolled in a transaction");
+        lock.lock();
+        try {
+            if (transactionContext != null) {
+                throw new SQLException("Auto-commit can not be set while enrolled in a transaction");
+            }
+            super.setAutoCommit(autoCommit);
+        } finally {
+            lock.unlock();
         }
-        super.setAutoCommit(autoCommit);
     }
 
 
     @Override
     public void commit() throws SQLException {
-        if (transactionContext != null) {
-            throw new SQLException("Commit can not be set while enrolled in a transaction");
+        lock.lock();
+        try {
+            if (transactionContext != null) {
+                throw new SQLException("Commit can not be set while enrolled in a transaction");
+            }
+            super.commit();
+        } finally {
+            lock.unlock();
```

Comment:
why we need to lock here ? it looks like just reading the transactionContext but not modifying it.

Reply:
Could it be unset by the reaper thread and would that result in the super.commit being allowed through? I think it might be the case?

Reply:
yeah, I understand. The reaper thread unset the transactionContext in the afterCompletion(int i) which could mean that the transaction is completed. So the super.commit is allowed when the connection is out of the transaction. I don't think we need the lock here.

Reply:
well, I think it is necessary because although the transaction is completed the application thread can still be associate with the transaction. However, I have changed this in my new PR: https://github.com/zhfeng/commons-dbcp/pull/2/files